pkg/gce-pd-csi-driver/ops_manager.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.checkAndUpdateLastKnownInstanceOps(ctx, project, instanceZone, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	op, err := o.cloudProvider.StartAttachDiskOp(ctx, volKey, readWrite, diskType, project, instanceZone, instanceName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("%v", err))
+	}
+
+	klog.V(5).Infof("AttachDisk operation %q for disk %q started on instance %q", op.Name, deviceName, instanceName)
+	o.opsCache.DiskInstanceOps.AddOp(common.CreateDiskInstanceKey(project, instanceZone, deviceName, instanceName), common.OpInfo{Name: op.Name, Type: op.OperationType})
+	return op, nil
+}
+
+// CheckCacheAndStartDetachDiskOp first verifies that there is no ongoing attach/detach operation for the given disk + instance combination, before triggering a new detach operation.
+func (o *OpsManager) checkCacheAndStartDetachDiskOp(ctx context.Context, project, instanceZone, deviceName, instanceName string) (*compute.Operation, error) {
+	o.Lock()
+	defer o.Unlock()
+
+	err := o.checkAndUpdateLastKnownDiskInstanceOp(ctx, project, instanceZone, deviceName, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = o.checkAndUpdateLastKnownInstanceOps(ctx, project, instanceZone, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	op, err := o.cloudProvider.StartDetachDiskOp(ctx, project, instanceZone, deviceName, instanceName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("%v", err))
+	}
+
+	klog.V(5).Infof("DetachDisk operation %q for disk %q started on instance %q", op.Name, deviceName, instanceName)
+	o.opsCache.DiskInstanceOps.AddOp(common.CreateDiskInstanceKey(project, instanceZone, deviceName, instanceName), common.OpInfo{Name: op.Name, Type: op.OperationType})
+	return op, nil
+}
+func (o *OpsManager) HydrateOpsCache() {
+	var ops []*compute.Operation
+	retry.OnError(backoff, func(err error) bool { return true }, func() error {
+		var err error
+		ops, err = o.cloudProvider.ListZonalOps(context.Background(), map[string]bool{
+			"attachDisk": true,
+			"detachDisk": true})
+		return err
+	})
+
+	klog.V(5).Infof("Found %d zonal ops", len(ops))
+	o.Lock()
+	defer o.Unlock()
+
+	for _, op := range ops {
+		if done, _ := o.cloudProvider.OpIsDone(op); done {
+			continue
+		}
+		project, zone, instanceName, err := common.ParseOpTargetLinkUrl(op.TargetLink)
+		if err != nil {
+			klog.Errorf("Failed to parse operation target link, err: %s", err)
+			continue
+		}
+
+		klog.V(5).Infof("Adding op %q (type %q) for project %q, zone %q, instance %q to cache", op.Name, op.OperationType, project, zone, instanceName)
+		o.opsCache.InstanceOps.AddOp(common.CreateInstanceKey(project, zone, instanceName), common.OpInfo{Name: op.Name, Type: op.OperationType})
+	}
+
+	o.ready = true
+}
+
+func (o *OpsManager) ClearDiskInstanceOpSafe(project, location, diskName, instanceName, opName string) error {
```

Comment:
Can you make this method lowercase? Now that the only callers are in this package, I don't think it needs to be exported

Reply:
done